Gradle build with KSP fails with "circular dependency" after Quarkus 2.13.0

### Describe the bug

When using KSP (Kotlin Symbol Processing, see https://kotlinlang.org/docs/ksp-overview.html), the Gradle build fails with a "circular dependency" starting with Quarkus 2.13.0 (up to the latest 2.14.2). It worked fine in 2.12.3.

### Expected behavior

Build succeeds

### Actual behavior

Build fails with circular dependency:

```
FAILURE: Build failed with an exception.

* What went wrong:
Circular dependency between the following tasks:
:backend:kspKotlin
+--- :backend:quarkusGenerateCode
|    \--- :backend:processResources
|         \--- :backend:kspKotlin (*)
\--- :backend:quarkusGenerateCodeDev
     \--- :backend:processResources (*)

(*) - details omitted (listed previously)
```

### How to Reproduce?

Reproducer: [quarkus-reproducer.zip](https://github.com/quarkusio/quarkus/files/10161870/quarkus-reproducer.zip)

Steps to reproduce:
1. `./gradlew clean build` (will fail)
2. In `gradle.properties`, change `quarkusVersion` from `2.13.0.Final` to `2.12.3.Final`.
3. `./gradlew clean build` (will succeed)

If `quarkusVersion` is greater or equal to 2.13.0.Final, it will fail in the same way, including the latest version (2.14.2.Final).

### Output of `uname -a` or `ver`

Darwin C02YK01PLVCG-ML 21.6.0 Darwin Kernel Version 21.6.0: Mon Aug 22 20:17:10 PDT 2022; root:xnu-8020.140.49~2/RELEASE_X86_64 x86_64

### Output of `java -version`

openjdk version "17.0.3" 2022-04-19

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

2.13.0

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Gradle 7.6

### Additional information

This occurs in at least Gradle 7.5.1 and 7.6.

Still seeing this issue on 2.16.3.

@Augustyn is your workaround only applicable to multi-module projects? Trying something similar on a smaller project I'm getting a big mess of Gradle errors:

```
FAILURE: Build failed with an exception.

* What went wrong:
Some problems were found with the configuration of task ':compileTestKotlin' (type 'KotlinCompile').
  - Gradle detected a problem with the following location: 'C:\projects\build\classes\java\quarkus-test-generated-sources'.
    
    Reason: Task ':compileTestKotlin' uses this output of task ':compileQuarkusTestGeneratedSourcesJava' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed.
    
    Possible solutions:
      1. Declare task ':compileQuarkusTestGeneratedSourcesJava' as an input of ':compileTestKotlin'.
      2. Declare an explicit dependency on ':compileQuarkusTestGeneratedSourcesJava' from ':compileTestKotlin' using Task#dependsOn.
      3. Declare an explicit dependency on ':compileQuarkusTestGeneratedSourcesJava' from ':compileTestKotlin' using Task#mustRunAfter.
    
    Please refer to https://docs.gradle.org/8.0.1/userguide/validation_problems.html#implicit_dependency for more details about this problem.
  - Gradle detected a problem with the following location: 'C:\projects\build\classes\java\quarkus-test-generated-sources\grpc'.
    
    Reason: Task ':compileTestKotlin' uses this output of task ':compileQuarkusTestGeneratedSourcesJava' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed.
    
    Possible solutions:
      1. Declare task ':compileQuarkusTestGeneratedSourcesJava' as an input of ':compileTestKotlin'.
      2. Declare an explicit dependency on ':compileQuarkusTestGeneratedSourcesJava' from ':compileTestKotlin' using Task#dependsOn.
      3. Declare an explicit dependency on ':compileQuarkusTestGeneratedSourcesJava' from ':compileTestKotlin' using Task#mustRunAfter.
    
    Please refer to https://docs.gradle.org/8.0.1/userguide/validation_problems.html#implicit_dependency for more details about this problem.
  - Gradle detected a problem with the following location: 'C:\projects\build\classes\java\quarkus-test-generated-sources\avdl'.
    
    Reason: Task ':compileTestKotlin' uses this output of task ':compileQuarkusTestGeneratedSourcesJava' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed.
    
    Possible solutions:
      1. Declare task ':compileQuarkusTestGeneratedSourcesJava' as an input of ':compileTestKotlin'.
      2. Declare an explicit dependency on ':compileQuarkusTestGeneratedSourcesJava' from ':compileTestKotlin' using Task#dependsOn.
      3. Declare an explicit dependency on ':compileQuarkusTestGeneratedSourcesJava' from ':compileTestKotlin' using Task#mustRunAfter.
    
    Please refer to https://docs.gradle.org/8.0.1/userguide/validation_problems.html#implicit_dependency for more details about this problem.
  - Gradle detected a problem with the following location: 'C:\projects\build\classes\java\quarkus-test-generated-sources\avpr'.
    
    Reason: Task ':compileTestKotlin' uses this output of task ':compileQuarkusTestGeneratedSourcesJava' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed.
    
    Possible solutions:
      1. Declare task ':compileQuarkusTestGeneratedSourcesJava' as an input of ':compileTestKotlin'.
      2. Declare an explicit dependency on ':compileQuarkusTestGeneratedSourcesJava' from ':compileTestKotlin' using Task#dependsOn.
      3. Declare an explicit dependency on ':compileQuarkusTestGeneratedSourcesJava' from ':compileTestKotlin' using Task#mustRunAfter.
    
    Please refer to https://docs.gradle.org/8.0.1/userguide/validation_problems.html#implicit_dependency for more details about this problem.
  - Gradle detected a problem with the following location: 'C:\projects\build\classes\java\quarkus-test-generated-sources\avsc'.
    
    Reason: Task ':compileTestKotlin' uses this output of task ':compileQuarkusTestGeneratedSourcesJava' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed.
    
    Possible solutions:
      1. Declare task ':compileQuarkusTestGeneratedSourcesJava' as an input of ':compileTestKotlin'.
      2. Declare an explicit dependency on ':compileQuarkusTestGeneratedSourcesJava' from ':compileTestKotlin' using Task#dependsOn.
      3. Declare an explicit dependency on ':compileQuarkusTestGeneratedSourcesJava' from ':compileTestKotlin' using Task#mustRunAfter.
```

> Still seeing this issue on 2.16.3.
> 
> @Augustyn is your workaround only applicable to multi-module projects? Trying something similar on a smaller project I'm getting a big mess of Gradle errors:
> 
> ```
> FAILURE: Build failed with an exception.
> 
> * What went wrong:
> Some problems were found with the configuration of task ':compileTestKotlin' (type 'KotlinCompile').
>   - Gradle detected a problem with the following location: 'C:\projects\build\classes\java\quarkus-test-generated-sources'.
>     
>     Reason: Task ':compileTestKotlin' uses this output of task ':compileQuarkusTestGeneratedSourcesJava' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed.
>     
>     Possible solutions:
>       1. Declare task ':compileQuarkusTestGeneratedSourcesJava' as an input of ':compileTestKotlin'.
>       2. Declare an explicit dependency on ':compileQuarkusTestGeneratedSourcesJava' from ':compileTestKotlin' using Task#dependsOn.
>       3. Declare an explicit dependency on ':compileQuarkusTestGeneratedSourcesJava' from ':compileTestKotlin' using Task#mustRunAfter.
>     
>     Please refer to https://docs.gradle.org/8.0.1/userguide/validation_problems.html#implicit_dependency for more details about this problem.
>   - Gradle detected a problem with the following location: 'C:\projects\build\classes\java\quarkus-test-generated-sources\grpc'.
>     
>     Reason: Task ':compileTestKotlin' uses this output of task ':compileQuarkusTestGeneratedSourcesJava' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed.
>     
>     Possible solutions:
>       1. Declare task ':compileQuarkusTestGeneratedSourcesJava' as an input of ':compileTestKotlin'.
>       2. Declare an explicit dependency on ':compileQuarkusTestGeneratedSourcesJava' from ':compileTestKotlin' using Task#dependsOn.
>       3. Declare an explicit dependency on ':compileQuarkusTestGeneratedSourcesJava' from ':compileTestKotlin' using Task#mustRunAfter.
>     
>     Please refer to https://docs.gradle.org/8.0.1/userguide/validation_problems.html#implicit_dependency for more details about this problem.
>   - Gradle detected a problem with the following location: 'C:\projects\build\classes\java\quarkus-test-generated-sources\avdl'.
>     
>     Reason: Task ':compileTestKotlin' uses this output of task ':compileQuarkusTestGeneratedSourcesJava' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed.
>     
>     Possible solutions:
>       1. Declare task ':compileQuarkusTestGeneratedSourcesJava' as an input of ':compileTestKotlin'.
>       2. Declare an explicit dependency on ':compileQuarkusTestGeneratedSourcesJava' from ':compileTestKotlin' using Task#dependsOn.
>       3. Declare an explicit dependency on ':compileQuarkusTestGeneratedSourcesJava' from ':compileTestKotlin' using Task#mustRunAfter.
>     
>     Please refer to https://docs.gradle.org/8.0.1/userguide/validation_problems.html#implicit_dependency for more details about this problem.
>   - Gradle detected a problem with the following location: 'C:\projects\build\classes\java\quarkus-test-generated-sources\avpr'.
>     
>     Reason: Task ':compileTestKotlin' uses this output of task ':compileQuarkusTestGeneratedSourcesJava' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed.
>     
>     Possible solutions:
>       1. Declare task ':compileQuarkusTestGeneratedSourcesJava' as an input of ':compileTestKotlin'.
>       2. Declare an explicit dependency on ':compileQuarkusTestGeneratedSourcesJava' from ':compileTestKotlin' using Task#dependsOn.
>       3. Declare an explicit dependency on ':compileQuarkusTestGeneratedSourcesJava' from ':compileTestKotlin' using Task#mustRunAfter.
>     
>     Please refer to https://docs.gradle.org/8.0.1/userguide/validation_problems.html#implicit_dependency for more details about this problem.
>   - Gradle detected a problem with the following location: 'C:\projects\build\classes\java\quarkus-test-generated-sources\avsc'.
>     
>     Reason: Task ':compileTestKotlin' uses this output of task ':compileQuarkusTestGeneratedSourcesJava' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed.
>     
>     Possible solutions:
>       1. Declare task ':compileQuarkusTestGeneratedSourcesJava' as an input of ':compileTestKotlin'.
>       2. Declare an explicit dependency on ':compileQuarkusTestGeneratedSourcesJava' from ':compileTestKotlin' using Task#dependsOn.
>       3. Declare an explicit dependency on ':compileQuarkusTestGeneratedSourcesJava' from ':compileTestKotlin' using Task#mustRunAfter.
> ```

Yes, his code was applicable to multi-module projects but I believe he also shared a snippet for single module projects:
```
project.afterEvaluate {
    getTasksByName("quarkusGenerateCode", true).forEach { task ->
        task.setDependsOn(task.dependsOn.map { it as Provider<Task> }.filter { it.get().name != "processResources" })
    }
    getTasksByName("quarkusGenerateCodeDev", true).forEach { task ->
        task.setDependsOn(task.dependsOn.map { it as Provider<Task> }.filter { it.get().name != "processResources" })
    }
}
```

--- a/build-parent/pom.xml
+++ b/build-parent/pom.xml
@@ -21,6 +21,7 @@
         <!-- These properties are needed in order for them to be resolvable by the generated projects -->
         <compiler-plugin.version>${version.compiler.plugin}</compiler-plugin.version>
         <kotlin.version>2.2.10</kotlin.version>
+        <ksp-gradle-plugin.version>2.2.10-2.0.2</ksp-gradle-plugin.version> <!-- Kotlin Symbol Processing plugin used in a test -->
         <dokka.version>2.0.0</dokka.version>
         <scala.version>2.13.12</scala.version>
         <scala-maven-plugin.version>4.9.5</scala-maven-plugin.version>

--- a/devtools/gradle/gradle-application-plugin/src/main/java/io/quarkus/gradle/QuarkusPlugin.java
+++ b/devtools/gradle/gradle-application-plugin/src/main/java/io/quarkus/gradle/QuarkusPlugin.java
@@ -7,6 +7,7 @@ import java.io.File;
 import java.nio.file.Path;
 import java.util.HashMap;
 import java.util.HashSet;
+import java.util.List;
 import java.util.Map;
 import java.util.Optional;
 import java.util.Set;
@@ -348,30 +349,21 @@ public class QuarkusPlugin implements Plugin<Project> {
 
                     project.afterEvaluate(this::afterEvaluate);
 
-                    tasks.named(JavaPlugin.COMPILE_JAVA_TASK_NAME, JavaCompile.class,
-                            compileJava -> {
-                                compileJava.mustRunAfter(quarkusGenerateCode);
-                                compileJava.mustRunAfter(quarkusGenerateCodeDev);
-                            });
-                    tasks.named(JavaPlugin.COMPILE_TEST_JAVA_TASK_NAME, JavaCompile.class,
-                            compileTestJava -> {
-                                compileTestJava.dependsOn(quarkusGenerateCode);
-                                compileTestJava.dependsOn(quarkusGenerateCodeTests);
-                                if (project.getGradle().getStartParameter().getTaskNames().contains(QUARKUS_DEV_TASK_NAME)) {
-                                    compileTestJava.getOptions().setFailOnError(false);
-                                }
-                            });
-
                     TaskProvider<Task> classesTask = tasks.named(JavaPlugin.CLASSES_TASK_NAME);
                     TaskProvider<Task> resourcesTask = tasks.named(JavaPlugin.PROCESS_RESOURCES_TASK_NAME);
                     TaskProvider<Task> testClassesTask = tasks.named(JavaPlugin.TEST_CLASSES_TASK_NAME);
                     TaskProvider<Task> testResourcesTask = tasks.named(JavaPlugin.PROCESS_TEST_RESOURCES_TASK_NAME);
+
+                    SourceSetContainer sourceSets = project.getExtensions().getByType(SourceSetContainer.class);
+                    SourceSet mainSourceSet = sourceSets.getByName(SourceSet.MAIN_SOURCE_SET_NAME);
+                    SourceSet testSourceSet = sourceSets.getByName(SourceSet.TEST_SOURCE_SET_NAME);
 
                     quarkusGenerateCode.configure(task -> {
                         Configuration config = project.getConfigurations().getByName(
                                 ApplicationDeploymentClasspathBuilder.getBaseRuntimeConfigName(LaunchMode.NORMAL));
                         task.dependsOn(resourcesTask, config);
                         task.setCompileClasspath(config);
+                        task.setSourcesDirectories(getSourcesParents(mainSourceSet));
                     });
                     quarkusGenerateCodeDev.configure(task -> {
                         Configuration config = project.getConfigurations().getByName(
@@ -379,12 +371,14 @@ public class QuarkusPlugin implements Plugin<Project> {
                                         .getBaseRuntimeConfigName(LaunchMode.DEVELOPMENT));
                         task.dependsOn(resourcesTask, config);
                         task.setCompileClasspath(config);
+                        task.setSourcesDirectories(getSourcesParents(mainSourceSet));
                     });
                     quarkusGenerateCodeTests.configure(task -> {
                         Configuration config = project.getConfigurations().getByName(
                                 ApplicationDeploymentClasspathBuilder.getBaseRuntimeConfigName(LaunchMode.TEST));
                         task.dependsOn(resourcesTask, config);
                         task.setCompileClasspath(config);
+                        task.setSourcesDirectories(getSourcesParents(testSourceSet));
                     });
 
                     quarkusDev.configure(task -> {
@@ -402,18 +396,6 @@ public class QuarkusPlugin implements Plugin<Project> {
                     });
                     quarkusBuildCacheableAppParts.configure(
                             task -> task.dependsOn(classesTask, resourcesTask, tasks.named(JavaPlugin.JAR_TASK_NAME)));
-
-                    SourceSetContainer sourceSets = project.getExtensions().getByType(SourceSetContainer.class);
-
-                    SourceSet mainSourceSet = sourceSets.getByName(SourceSet.MAIN_SOURCE_SET_NAME);
-                    SourceSet testSourceSet = sourceSets.getByName(SourceSet.TEST_SOURCE_SET_NAME);
-
-                    mainSourceSet.getJava().srcDirs(quarkusGenerateCode, quarkusGenerateCodeDev);
-                    testSourceSet.getJava().srcDirs(quarkusGenerateCodeTests);
-
-                    quarkusGenerateCode.configure(task -> task.setSourcesDirectories(getSourcesParents(mainSourceSet)));
-                    quarkusGenerateCodeDev.configure(task -> task.setSourcesDirectories(getSourcesParents(mainSourceSet)));
-                    quarkusGenerateCodeTests.configure(task -> task.setSourcesDirectories(getSourcesParents(testSourceSet)));
 
                     SourceSet intTestSourceSet = sourceSets.getByName(INTEGRATION_TEST_SOURCE_SET_NAME);
                     intTestSourceSet.setCompileClasspath(
@@ -487,20 +469,27 @@ public class QuarkusPlugin implements Plugin<Project> {
                     // quarkusBuild is expected to run after the project has passed the tests
                     quarkusBuildCacheableAppParts.configure(task -> task.shouldRunAfter(tasks.withType(Test.class)));
 
-                    SourceSet generatedSourceSet = sourceSets.getByName(QuarkusGenerateCode.QUARKUS_GENERATED_SOURCES);
-                    SourceSet generatedTestSourceSet = sourceSets.getByName(QuarkusGenerateCode.QUARKUS_TEST_GENERATED_SOURCES);
-
-                    project.afterEvaluate(project1 -> {
-                        // Register the quarkus-generated-code
-                        for (String provider : quarkusExt.getCodeGenerationProviders().get()) {
-
-                            mainSourceSet.getJava().srcDir(
-                                    new File(generatedSourceSet.getJava().getClassesDirectory().get().getAsFile(), provider));
-                            testSourceSet.getJava().srcDir(
-                                    new File(generatedTestSourceSet.getJava().getClassesDirectory().get().getAsFile(),
-                                            provider));
-                        }
-                    });
+                    tasks.named(JavaPlugin.COMPILE_JAVA_TASK_NAME, JavaCompile.class,
+                            compileJava -> {
+                                // quarkusGenerateCode is a dependency
+                                compileJava.dependsOn(quarkusGenerateCode);
+                                // quarkusGenerateCodeDev must run before compileJava in case quarkusDev is the target
+                                compileJava.mustRunAfter(quarkusGenerateCodeDev);
+                                // add the code gen sources
+                                addCodeGenSourceDirs(compileJava,
+                                        sourceSets.getByName(QuarkusGenerateCode.QUARKUS_GENERATED_SOURCES), quarkusExt);
+                            });
+                    tasks.named(JavaPlugin.COMPILE_TEST_JAVA_TASK_NAME, JavaCompile.class,
+                            compileTestJava -> {
+                                compileTestJava.dependsOn(quarkusGenerateCode);
+                                compileTestJava.dependsOn(quarkusGenerateCodeTests);
+                                if (project.getGradle().getStartParameter().getTaskNames().contains(QUARKUS_DEV_TASK_NAME)) {
+                                    compileTestJava.getOptions().setFailOnError(false);
+                                }
+                                // add the code gen test sources
+                                addCodeGenSourceDirs(compileTestJava,
+                                        sourceSets.getByName(QuarkusGenerateCode.QUARKUS_TEST_GENERATED_SOURCES), quarkusExt);
+                            });
                 });
 
         project.getPlugins().withId("org.jetbrains.kotlin.jvm", plugin -> {
@@ -508,9 +497,39 @@ public class QuarkusPlugin implements Plugin<Project> {
             tasks.named("compileKotlin", task -> {
                 task.mustRunAfter(quarkusGenerateCode);
                 task.mustRunAfter(quarkusGenerateCodeDev);
+                addCodeGenSourceDirs(task, project.getExtensions().getByType(SourceSetContainer.class)
+                        .getByName(QuarkusGenerateCode.QUARKUS_GENERATED_SOURCES), quarkusExt);
             });
-            tasks.named("compileTestKotlin", task -> task.dependsOn(quarkusGenerateCodeTests));
+            tasks.named("compileTestKotlin", task -> {
+                task.dependsOn(quarkusGenerateCodeTests);
+                addCodeGenSourceDirs(task, project.getExtensions().getByType(SourceSetContainer.class)
+                        .getByName(QuarkusGenerateCode.QUARKUS_TEST_GENERATED_SOURCES), quarkusExt);
+            });
         });
+    }
+
+    private static void addCodeGenSourceDirs(JavaCompile compileJava, SourceSet generatedSourceSet,
+            QuarkusPluginExtension quarkusExt) {
+        final File baseDir = generatedSourceSet.getJava().getClassesDirectory().get().getAsFile();
+        for (String provider : quarkusExt.getCodeGenerationProviders().get()) {
+            compileJava.source(new File(baseDir, provider));
+        }
+    }
+
+    private static void addCodeGenSourceDirs(Task compileKotlin, SourceSet generatedSourceSet,
+            QuarkusPluginExtension quarkusExt) {
+        final File baseDir = generatedSourceSet.getJava().getClassesDirectory().get().getAsFile();
+        final List<String> codeGenProviders = quarkusExt.getCodeGenerationProviders().get();
+        final Object[] codeGenDirs = new Object[codeGenProviders.size()];
+        for (int i = 0; i < codeGenDirs.length; ++i) {
+            codeGenDirs[i] = new File(baseDir, codeGenProviders.get(i));
+        }
+        try {
+            var sourcesMethod = compileKotlin.getClass().getMethod("source", Object[].class);
+            sourcesMethod.invoke(compileKotlin, new Object[] { codeGenDirs });
+        } catch (Exception e) {
+            throw new RuntimeException(e);
+        }
     }
 
     private ApplicationDeploymentClasspathBuilder getDeploymentClasspathBuilder(Project project, LaunchMode mode) {

--- a/devtools/gradle/gradle-application-plugin/src/test/java/io/quarkus/gradle/tasks/CachingTest.java
+++ b/devtools/gradle/gradle-application-plugin/src/test/java/io/quarkus/gradle/tasks/CachingTest.java
@@ -40,7 +40,6 @@ import org.junit.jupiter.params.provider.MethodSource;
 public class CachingTest {
     private static final Map<String, TaskOutcome> ALL_SUCCESS = Map.of(
             ":quarkusGenerateCode", TaskOutcome.SUCCESS,
-            ":quarkusGenerateCodeDev", TaskOutcome.SUCCESS,
             ":quarkusGenerateCodeTests", TaskOutcome.SUCCESS,
             ":quarkusAppPartsBuild", TaskOutcome.SUCCESS,
             ":quarkusDependenciesBuild", TaskOutcome.SUCCESS,
@@ -56,7 +55,6 @@ public class CachingTest {
             ":build", TaskOutcome.UP_TO_DATE);
     public static final Map<String, TaskOutcome> FROM_CACHE = Map.of(
             ":quarkusGenerateCode", TaskOutcome.FROM_CACHE,
-            ":quarkusGenerateCodeDev", TaskOutcome.SUCCESS,
             ":quarkusGenerateCodeTests", TaskOutcome.FROM_CACHE,
             ":quarkusAppPartsBuild", TaskOutcome.FROM_CACHE,
             ":quarkusDependenciesBuild", TaskOutcome.SUCCESS,
@@ -211,7 +209,7 @@ public class CachingTest {
                 .describedAs("output: %s", result.getOutput())
                 .containsEntry(":compileJava", TaskOutcome.FROM_CACHE)
                 .containsEntry(":quarkusGenerateCode", TaskOutcome.FROM_CACHE)
-                .containsEntry(":quarkusGenerateCodeDev", TaskOutcome.UP_TO_DATE)
+                .doesNotContainKey(":quarkusGenerateCodeDev")
                 .containsEntry(":quarkusAppPartsBuild", isFastOrLegacyJar ? TaskOutcome.FROM_CACHE : TaskOutcome.UP_TO_DATE)
                 .containsEntry(":quarkusDependenciesBuild", isFastOrLegacyJar ? TaskOutcome.SUCCESS : TaskOutcome.UP_TO_DATE)
                 .containsEntry(":quarkusBuild", simulateCI || isFastJar ? TaskOutcome.SUCCESS : TaskOutcome.FROM_CACHE);

--- a/devtools/gradle/gradle-application-plugin/src/test/java/io/quarkus/gradle/tasks/EagerResolutionTaskTest.java
+++ b/devtools/gradle/gradle-application-plugin/src/test/java/io/quarkus/gradle/tasks/EagerResolutionTaskTest.java
@@ -46,7 +46,7 @@ public class EagerResolutionTaskTest {
         String addingQuarkusExtension = """
                 quarkus {
                   cachingRelevantProperties.add("FOO_ENV_VAR")
-                  quarkusBuildProperties.put("quarkus.package.type", "fast-jar")
+                  quarkusBuildProperties.put("quarkus.package.jar.type", "fast-jar")
                   quarkusBuildProperties.putAll(
                     provider {
                       tasks

--- a/integration-tests/gradle/src/main/resources/kotlin-ksp/build.gradle.kts
+++ b/integration-tests/gradle/src/main/resources/kotlin-ksp/build.gradle.kts
@@ -1,0 +1,47 @@
+plugins {
+    kotlin("jvm")
+    kotlin("plugin.allopen")
+    id("io.quarkus")
+    id("com.google.devtools.ksp")
+}
+
+repositories {
+    mavenCentral()
+    mavenLocal()
+}
+
+val quarkusPlatformGroupId: String by project
+val quarkusPlatformArtifactId: String by project
+val quarkusPlatformVersion: String by project
+
+dependencies {
+    implementation(enforcedPlatform("${quarkusPlatformGroupId}:${quarkusPlatformArtifactId}:${quarkusPlatformVersion}"))
+    implementation("io.quarkus:quarkus-kotlin")
+    implementation("org.jetbrains.kotlin:kotlin-stdlib-jdk8")
+    implementation("io.quarkus:quarkus-rest")
+    testImplementation("io.quarkus:quarkus-junit5")
+    testImplementation("io.rest-assured:rest-assured")
+}
+
+group = "org.acme"
+version = "1.0.0-SNAPSHOT"
+
+tasks.withType<Test> {
+    systemProperty("java.util.logging.manager", "org.jboss.logmanager.LogManager")
+}
+allOpen {
+    annotation("jakarta.ws.rs.Path")
+    annotation("jakarta.enterprise.context.ApplicationScoped")
+    annotation("io.quarkus.test.junit.QuarkusTest")
+}
+
+kotlin {
+    compilerOptions {
+        javaParameters = true
+    }
+}
+
+java {
+    // test that the sources JAR works with the KSP plugin
+    withSourcesJar()
+}

--- a/integration-tests/gradle/src/main/resources/kotlin-ksp/gradle.properties
+++ b/integration-tests/gradle/src/main/resources/kotlin-ksp/gradle.properties
@@ -1,0 +1,5 @@
+quarkusPluginId=io.quarkus
+quarkusPlatformGroupId=io.quarkus
+quarkusPlatformArtifactId=quarkus-bom
+kotlinVersion=${kotlin.version}
+kspVersion=${ksp-gradle-plugin.version}

--- a/integration-tests/gradle/src/main/resources/kotlin-ksp/settings.gradle.kts
+++ b/integration-tests/gradle/src/main/resources/kotlin-ksp/settings.gradle.kts
@@ -1,0 +1,23 @@
+pluginManagement {
+    val quarkusPluginVersion: String by settings
+    val quarkusPluginId: String by settings
+    val kotlinVersion: String by settings
+    val kspVersion: String by settings
+    repositories {
+        mavenLocal {
+            content {
+                includeGroupByRegex("io.quarkus.*")
+                includeGroup("org.hibernate.orm")
+            }
+        }
+        mavenCentral()
+        gradlePluginPortal()
+    }
+    plugins {
+        id(quarkusPluginId) version quarkusPluginVersion
+        kotlin("jvm") version kotlinVersion
+        kotlin("plugin.allopen") version kotlinVersion
+        id("com.google.devtools.ksp") version kspVersion
+    }
+}
+rootProject.name="code-with-quarkus"

--- a/integration-tests/gradle/src/main/resources/kotlin-ksp/src/main/kotlin/org/acme/GreetingResource.kt
+++ b/integration-tests/gradle/src/main/resources/kotlin-ksp/src/main/kotlin/org/acme/GreetingResource.kt
@@ -1,0 +1,14 @@
+package org.acme
+
+import jakarta.ws.rs.GET
+import jakarta.ws.rs.Path
+import jakarta.ws.rs.Produces
+import jakarta.ws.rs.core.MediaType
+
+@Path("/hello")
+class GreetingResource {
+
+    @GET
+    @Produces(MediaType.TEXT_PLAIN)
+    fun hello() = "Hello from Quarkus REST"
+}

--- a/integration-tests/gradle/src/main/resources/kotlin-ksp/src/test/kotlin/org/acme/GreetingResourceTest.kt
+++ b/integration-tests/gradle/src/main/resources/kotlin-ksp/src/test/kotlin/org/acme/GreetingResourceTest.kt
@@ -1,0 +1,20 @@
+package org.acme
+
+import io.quarkus.test.junit.QuarkusTest
+import io.restassured.RestAssured.given
+import org.hamcrest.CoreMatchers.`is`
+import org.junit.jupiter.api.Test
+
+@QuarkusTest
+class GreetingResourceTest {
+
+    @Test
+    fun testHelloEndpoint() {
+        given()
+          .`when`().get("/hello")
+          .then()
+             .statusCode(200)
+             .body(`is`("Hello from Quarkus REST"))
+    }
+
+}

--- a/integration-tests/gradle/src/test/java/io/quarkus/gradle/KspPluginWithSourcesJarTest.java
+++ b/integration-tests/gradle/src/test/java/io/quarkus/gradle/KspPluginWithSourcesJarTest.java
@@ -1,0 +1,18 @@
+package io.quarkus.gradle;
+
+import java.io.File;
+
+import org.junit.jupiter.api.Test;
+
+/**
+ * Tests compatibility of the KSP plugin with sourcesJar task.
+ * Previously, the Quarkus plugin added the codegen sources to the main sources and that created a cyclic task dependencies.
+ */
+public class KspPluginWithSourcesJarTest extends QuarkusGradleWrapperTestBase {
+
+    @Test
+    public void testKspWithSourcesJar() throws Exception {
+        final File projectDir = getProjectDir("kotlin-ksp");
+        runGradleWrapper(projectDir, "clean", "build");
+    }
+}

